### ISSUE TITLE
default TEMPLATE_PATH error

### ISSUE THREAD

**Issue by qhsoft** [bottlepy/bottle#1141]
spath = os.path.abspath(spath) + os.sep
need modify to:  spath = os.path.abspath(spath)
fname = os.path.abspath(os.path.join(spath, name))

**Comment by defnull** [comment]
Please describe why this is an error. The `+ os.path` seems to have no effect, but it also does not cause wrong behavior.

Pull requests welcomed.

**Comment by julenl** [comment]
I can confirm. Some windows machines get confused with the '\\\\' at the end of the path.
I worked around the issue by inserting the the "./views" (without last slash) folder at the beginning of the `bottle.TEMPLATE_PATH` variable in order to make it find the templates.

**Comment by defnull** [comment]
> I can confirm. Some windows machines get confused with the '\\' at the end of the path.

Ahh interesting! The `os.pth.join` does not remove duplicate separators then?

**Comment by julenl** [comment]
It is difficult to reproduce. In most of windows computers I have seen it works fine, but for some reason I just saw it not working on one: Windows 10, fresh install from Python 3.7.2 and latest bottle from pip.

**Comment by defnull** [comment]
`os.path.join` returns the same output with or without a leading slashes, but I can only test this on Linux. I suspect that on windows, it gets confused if you are mixing slashes and backslashes. 

Can someone please run this on windows 10?

```python
>>> for p1 in ("/foo", "/foo"+os.sep):
...   for p2 in ("bar", "./bar", "."+os.sep+"bar"):
...     print(p1, p2, os.path.join(p1, p2))
... 
('/foo', 'bar', '/foo/bar')
('/foo', './bar', '/foo/./bar')
('/foo', './bar', '/foo/./bar')
('/foo/', 'bar', '/foo/bar')
('/foo/', './bar', '/foo/./bar')
('/foo/', './bar', '/foo/./bar')
```

**Comment by v-python** [comment]
On 5/28/2019 6:07 AM, Marcel Hellkamp wrote:
>
> |os.path.join| returns the same output with or without a leading 
> slashes, but I can only test this on Linux. I suspect that on windows, 
> it gets confused if you are mixing slashes and backslashes.
>
> Can someone please run this on windows 10?
>
> >>>  for  p1in  ("/foo","/foo"+os.sep):
> ...    for  p2in  ("bar","./bar","."+os.sep+"bar"):
> ...      print(p1, p2, os.path.join(p1, p2))
> ...  
> ('/foo','bar','/foo/bar')
> ('/foo','./bar','/foo/./bar')
> ('/foo','./bar','/foo/./bar')
> ('/foo/','bar','/foo/bar')
> ('/foo/','./bar','/foo/./bar')
> ('/foo/','./bar','/foo/./bar')
>
 >>> import os
 >>> for p1 in ("/foo", "/foo"+os.sep):
...     for p2 in ("bar", "./bar", "."+os.sep+"bar"):
...         print(p1, p2, os.path.join(p1, p2))
...
/foo bar /foo\bar
/foo ./bar /foo\./bar
/foo .\bar /foo\.\bar
/foo\ bar /foo\bar
/foo\ ./bar /foo\./bar
/foo\ .\bar /foo\.\bar
 >>>

**Comment by mcarpenter** [comment]
Hello,

I just hit this. I will provide below an explanation, proposed solution, diff, small test case and workaround that can be used until the next release.

### Explanation

I think the problem is not so much `os.path.join()` not removing duplicate _trailing_ (not leading) slashes but rather the differing behavior of `os.path.abspath()` (from which the arguments to `join()` are constructed) when it is supplied an argument with a trailing slash. Compare and contrast the following on the default values of `TEMPLATE_PATH = ['./', './views/']` on Linux and Windows. (It does not matter if directory `views` actually exists or not; results are the same on both platforms).

On Linux, from `$HOME`:
```
os.path.abspath('./') => '/home/mcarpenter'
os.path.abspath(./views') => '/home/mcarpenter/views'
```

On Windows, from `%HOME%`:
```
os.path.abspath('./') => 'C:\\Users\\mcarpenter\\'
os.path.abspath(./views') => 'C:\\Users\\mcarpenter\\views\\'
```

This is arguably a bug: it's certainly inconsistent.

Then after appending `os.sep` the template directory `spath` has either one (Linux) or two (Win) trailing slashes. 

Subsequently `os.path.join()` to make the proposed template path `fname` is a little peculiar (but identical on both platforms): if there is no trailing slash on the first component, then it introduces one (of course). If there are one or more trailing slashes on the first component, then they are (all) retained. That is, repeated trailing slashes on the first component are not coalesced.

Finally, `abspath()` is called on `fname` and this zaps the erroneous double slash. So on Win we have something like:

```python
spath = 'C:\\Users\\mcarpenter\\\\' # two trailing slashes
fname = 'C:\\Users\\mcarpenter\template_name'
```

Consequently `fname.startswith(spath)` is `False`, and `spath` is rejected.

### Solution

What I think this should be doing is:

1. Remove trailing `os.sep`s on the `TEMPLATE_PATH` components. This solves the current default template path and the case where consumers might specify one (or more!) trailing slashes.
2. Rely on `os.path.join()` to add precisely one separator between components.
3. Remove the `startswith()` test. I don't see a case where this should fail: `abspath()` is not a proof of existence (it's not `realpath()`) and we've just explicitly constructed the path to start with the lookup directory. So I think this is redundant?

(I assume template names never start with slashes; user-supplied template names are a -- hopefully obvious -- security risk).

### Diff

```diff
diff --git a/bottle.py b/bottle.py
index 96ca3e4..0d3ff9f 100644
--- a/bottle.py
+++ b/bottle.py
@@ -3247,9 +3247,8 @@ class BaseTemplate(object):
             return os.path.abspath(name)
 
         for spath in lookup:
-            spath = os.path.abspath(spath) + os.sep
+            spath = os.path.abspath(spath).rstrip(os.sep)
             fname = os.path.abspath(os.path.join(spath, name))
-            if not fname.startswith(spath): continue
             if os.path.isfile(fname): return fname
             for ext in cls.extensions:
                 if os.path.isfile('%s.%s' % (fname, ext)):
```

### Test

Very quick test now passes on both platforms:

`test.py`:
```python
from bottle import template
print(template('test_template'))
```

`test_template.tpl`:
```
test template
```

### Workaround

Before this is fixed, a workaround for Windows users is to override `bottle.TEMPLATE_PATH` with directories that do not have a trailing slash (also compatible with Linux):

```python
import bottle
...
bottle.TEMPLATE_PATH = ['.', 'views']
```

**Comment by mcarpenter** [comment]
I worked out why this was hard to track down: it was a Python-on-Windows regression with releases 3.6.7 and 3.7.1 (I have the latter).

 * https://bugs.python.org/issue35205
 * https://github.com/python/cpython/pull/10082

This has allegedly been fixed in subsequent versions of Python (3.6.10 and 3.7.6 are current).

**Comment by defnull** [comment]
Ah, that explains why I could not reproduce it. Thanks for the write-up and clarification! Guess we can close this now, then.
